packages/lisk-transactions/src/transactions/3_vote_transaction.ts

```diff
@@ -0,0 +1,483 @@
+
+		const errors = sameTypeTransactions.reduce(
+			(previous, tx) => {
+				if (
+					tx.asset.votes
+						.map(vote => vote.substring(1))
+						.some(publicKey => publicKeys.includes(publicKey))
+				) {
+					return [
+						...previous,
+						new TransactionError(
+							'Transaction includes conflicting vote',
+							tx.id,
+							'.asset.votes',
+						),
+					];
+				}
+
+				return previous;
+			},
+			[] as ReadonlyArray<TransactionError>,
+		);
+
+		return {
+			id: this.id,
+			errors,
+			status: errors.length === 0 ? Status.OK : Status.FAIL,
+		};
+	}
+
+	public processRequiredState(state: EntityMap): RequiredVoteState {
+		const accounts = state[ENTITY_ACCOUNT];
+		if (!accounts) {
+			throw new Error('Entity account is required.');
+		}
+		if (
+			!isTypedObjectArrayWithKeys<Account>(accounts, ['address', 'publicKey'])
+		) {
+			throw new Error('Required state does not have valid account type.');
+		}
+
+		const sender = accounts.find(acct => acct.address === this.senderId);
+		if (!sender) {
+			throw new Error('No sender account is found.');
+		}
+		const votes = this.asset.votes.map(vote => vote.substring(1));
+		const dependentAccounts = accounts.filter(acct =>
+			votes.includes(acct.publicKey),
+		);
+		if (votes.length !== dependentAccounts.length) {
+			throw new Error('Not enought accounts in dependent state.');
+		}
+
+		return {
+			sender,
+			dependentState: {
+				[ENTITY_ACCOUNT]: dependentAccounts,
+			},
+		};
+	}
+
+	public validateSchema(): TransactionResponse {
+		const { errors: baseErrors, status } = super.validateSchema();
+		const valid = validator.validate(voteAssetFormatSchema, this.asset);
+		const errors = [...baseErrors];
+		if (!this.amount.eq(0)) {
+			errors.push(
+				new TransactionError(
+					'Amount must be zero for vote transaction',
+					this.id,
+					'.asset',
+				),
+			);
+		}
+
+		if (
+			this.recipientPublicKey === undefined ||
+			this.recipientPublicKey === ''
+		) {
+			errors.push(
+				new TransactionError(
+					'RecipientId must be set for vote transaction',
+					this.id,
+					'.recipientPublicKey',
+				),
+			);
+		}
+
+		if (this.recipientId === undefined || this.recipientId === '') {
+			errors.push(
+				new TransactionError(
+					'RecipientId must be set for vote transaction',
+					this.id,
+					'.recipientId',
+				),
+			);
+		}
+
+		const assetErrors = validator.errors
+			? validator.errors.map(
+					error =>
+						new TransactionError(
+							`'${error.dataPath}' ${error.message}`,
+							this.id,
+							error.dataPath,
+						),
+			  )
+			: [];
+		errors.push(...assetErrors);
+
+		return {
+			id: this.id,
+			status:
+				status === Status.OK && valid && errors.length === 0
+					? Status.OK
+					: Status.FAIL,
+			errors,
+		};
+	}
+
+	public verify({
+		sender,
+		dependentState,
+	}: RequiredVoteState): TransactionResponse {
+		const { errors: baseErrors } = super.apply({ sender });
```

Comment:
verify will be called after `processRequiredState`. So, do we need to check here if the passed arguments match are as expected?

Reply:
typescript wise, it's hard to check without going through the basic check ='(